aea/agent_loop.py

```diff
@@ -0,0 +1,426 @@
+        if not self._timerhandle:
+            return
+
+        self._timerhandle.cancel()
+        self._timerhandle = None
+
+
+class BaseAgentLoop(ABC):
+    """Base abstract  agent loop class."""
+
+    def __init__(self, agent: "Agent") -> None:
+        """Init loop.
+
+        :params agent: Agent or AEA to run.
+        """
+        self._agent = agent
+
+    @abstractmethod
+    def start(self) -> None:
+        """Start agent loop."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def stop(self) -> None:
+        """Stop agent loop."""
+        raise NotImplementedError
+
+
+class AgentLoopException(AEAException):
+    """Exception for agent loop runtime errors."""
+
+
+class AgentLoopStates(Enum):
+    """Internal agent loop states."""
+
+    initial = None
+    started = "started"
+    stopped = "stopped"
+    stopping = "stopping"
+    error = "error"
+
+
+class AsyncAgentLoop(BaseAgentLoop):
+    """Asyncio based agent loop suitable only for AEA."""
+
+    NEW_BEHAVIOURS_PROCESS_SLEEP = 1  # check new behaviours registered every second.
+
+    def __init__(self, agent: "AEA", loop: AbstractEventLoop = None):
+        """
+        Init agent loop.
+
+        :param agent: AEA instance
+        :param loop: asyncio loop to use. optional
+        """
+        super().__init__(agent)
+        self._agent: "AEA" = self._agent
+
+        try:
+            self._loop = loop or asyncio.get_event_loop()
+            assert not self._loop.is_closed()
+            assert not self._loop.is_running()
+        except (RuntimeError, AssertionError):
+            self._loop = asyncio.new_event_loop()
+            asyncio.set_event_loop(self._loop)
+
+        self._behaviours_registry: Dict[Behaviour, PeriodicCaller] = {}
+        self._state: AsyncState = AsyncState()
+        self._exceptions: List[Exception] = []
+
+    def start(self):
+        """Start agent loop."""
+        self._state.state = AgentLoopStates.started
+        self._loop.run_until_complete(self._run())
+
+    def stop(self):
+        """Stop agent loop."""
+        self._state.state = AgentLoopStates.stopping
+
+    def _behaviour_exception_callback(self, fn: Callable, exc: Exception) -> None:
+        """
+        Call on behaviour's act exception.
+
+        :param fn: behaviour's act
+        :param: exc: Exception  raised
+
+        :return: None
+        """
+        logger.exception(f"Exception: `{exc}` occured during `{fn}` processing")
+        self._exceptions.append(exc)
+        self._state.state = AgentLoopStates.error
+
+    def _register_behaviour(self, behaviour: Behaviour) -> None:
+        """
+        Register behaviour to run periodically.
+
+        :param behaviour: Behaviour object
+
+        :return: None
+        """
+        if behaviour in self._behaviours_registry:
+            # already registered
+            return
+
+        periodic_caller = PeriodicCaller(
+            behaviour.act,
+            behaviour._tick_interval,
```

Comment:
with the current implementation, behaviour code is allowed to change its tick interval. I'm aware that this was the case also in the previous implementation. However, I wonder: is it our desired property?

Reply:
Moreover, do you see any problems in letting the user set `tick_interval` to `0.0`?
Maybe we could still read from `behaviour._tick_interval`, but we could do some check like:
```
max(behaviour._tick_interval, 0.001)
```

Reply:
runtime change of interval was discussed and decided that it should be immutable in runtime, but can check with David.

if user set tick interval to 0, or -1  or None or string -it's a user's problem.
otherwise we can make strict parameters check and raise value error

Reply:
>runtime change of interval was discussed and decided that it should be immutable in runtime, but can check with David.

However, it seems to me it is mutable because I can write something like:

```python

class MyBehaviour(Behaviour):

    def act(self):
        """Change tick interval at runtime"""
        self._tick_interval = self._tick_interval * 0.9

```

So the tick interval gets smaller and smaller, and if I deactivate and re-activate it, the new tick interval will be loaded. Anyway, it's a corner case, we can ignore it.